Add an AST Element Example Doc (#100)
openrewrite/rewrite-docs: reference/ast-examples.md

# AST Element Examples

> When building recipes, it's important to understand how certain elements in the OpenRewrite Abstract Syntax Tree (AST) correspond to code. You couldn't, for example, properly rename a variable with a recipe unless you knew that `J.Identifier` is how variable names are represented in the AST.

Reviewer: I note the term "element" is used extensively throughout this document, but "element" isn't a term we define anywhere.

I wonder if using this term repeatedly may be confusing. It feels like it implies some formal distinction or concept that isn't there. "not complete/total" or "component of" is what I usually mean by "element" when I say "AST element", but readers might be wondering if it means something more formal. Is a `J.CompilationUnit` an "element" of an AST since it is a whole AST?

Reviewer: Part of the language trickiness here is than an AST element is itself an AST. So in conversation we normally say "AST" to mean "the whole AST" and "AST element" to mean "any AST which is itself a component of another, larger AST"

Reviewer 2: This is an interesting distinction. I think I have been guilty in the past of calling them "elements of an AST". Are you suggesting replacing `element` with `ast`? I think this makes sense the more I think about it...

"AST Element Examples" could just be "AST Examples"..."under how certain ASTs in the ....." etc.

* [Code Sample](#code)
  * [Binary](#binary)
  * [Block](#block)
  * [ClassDeclaration](#classdeclaration)
  * [CompilationUnit](#compilationunit)
  * [Expression](#expression)
  * [FieldAccess](#fieldaccess)
  * [Identifier](#identifier)
  * [MethodDeclaration](#methoddeclaration)
  * [MethodInvocation](#methodinvocation)
  * [NewClass](#newclass)
  * [Statement](#statement)
  * [VariableDeclarations](#variabledeclarations)
* [Using the Debugger To Detect Elements](#using-the-debugger-to-detect-elements)

## Code

Below is a simple Java class whose entire purpose is to demonstrate different AST elements. Each of the following sections will highlight different parts of this code to demonstrate which chunks correspond to which AST element.

```java
package org.openrewrite;
                
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@AnAnnotation
public class A {

    List<Integer> a = new ArrayList<>();

    int foo() {
        int a = 1 + 2, b = 3, c;
        this.a = this.a.stream()
                .map(it -> it + 1)
                .collect(Collectors.toList());
        return a;
    }
}

@interface AnAnnotation {}
```

### Binary

> A [Binary element](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L597-L705) is any operation with a left and right side (usually separated by something like a `+`).

Reviewer: ```suggestion
A [Binary](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L597-L705) is an [Expression](#expression) with a left and right side, separated by an operator. Examples of operators include `+`, `-`, `||`, `&&`, and more.
```

![Binary Element Example](<../.gitbook/assets/Binary.png>)

### Block 

> A [Block element](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L712-L851) is the code contained between curly braces. Blocks can be nested inside of each other.

Reviewer: ```suggestion
A [Block](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L712-L851) is a pair of curly braces and the [Statements](#statement) contained within. Blocks can be nested inside of each other.
```

![Block Element Example](<../.gitbook/assets/Block.png>)

### ClassDeclaration

> A [ClassDeclaration element](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L1062-L1336) contains all of the code for any Java class. 

Reviewer 2: I think it would be a good idea to remind users that a ClassDeclaration can be nested inside another class: 

```
public class A {
...

   private class B {
   ...
   }
}
```

### Expression

An [Expression element](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/Expression.java) is anything that returns a value. `MethodInvocation`, `Identifier`, and `Binary` are all examples of expressions. Please note that some AST elements such as `MethodInvocation` are both a [Statement](#statement) and an Expression.

In the below code, only some of the expressions are highlighted as expressions can often have many expressions inside of them and it would be too difficult to read if all of them were highlighted. For instance, `import java.util.ArrayList` is many expressions (`java`, `util`, `ArrayList`, `java.util`, and `java.util.ArrayList`).

![Expression Element Example](<../.gitbook/assets/Expression.png>)

### FieldAccess

A [FieldAccess element](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L1827-L1953) is any fully qualified name. Often times, these are package or import statements, but they can also appear in code as something like: `this.foo`.

![FieldAccess Element Example](<../.gitbook/assets/FieldAccess.png>)

### Identifier

> An [Identifier element](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L2310-L2343) is any name in the code (class names, variable names, method names, etc).

Reviewer: Good place to mention that `J.Identifier.getFieldType()` can be used to tell what class the identifier is a field on. `null` for identifiers which are not fields.

![Identifier Element Example](<../.gitbook/assets/Identifier.png>)

### MethodDeclaration

> A [MethodDeclaration element](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L3517) is all the code for creating a method.

Reviewer: ```suggestion
A [MethodDeclaration](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L3517) is the annotations, modifiers, return type, name, argument list, and body which together define a method on a [class](#classdeclaration).
```

# AST Examples

> When building recipes, it's important to understand how OpenRewrite Abstract Syntax Trees (AST) correspond to code. You couldn't, for example, properly rename a variable with a recipe unless you knew that [J.Identifier](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L2310-L2343) is the class used to represent a variable.

Reviewer 3: Perfect! I missed that comment.

# AST Examples

When building recipes, it's important to understand how OpenRewrite Abstract Syntax Trees (AST) correspond to code. You couldn't, for example, properly rename a variable with a recipe unless you knew that [J.Identifier](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L2310-L2343) is the class used to represent a variable.

> To help you get started on working with ASTs and OpenRewrite, this guide will:

Reviewer: We should mention that this is a guide to the Java AST. XML, Protobuf, Yaml, etc., each have their own AST. Being data formats those are generally simpler and more approachable structures.

* [Explain how ASTs work at a high-level](#high-level-ast-explanation)
* [Provide a sample chunk of code to demonstrate ASTs](#code)
* Discuss how that code relates to different types of ASTs
  * [Binary](#binary)
  * [Block](#block)
  * [ClassDeclaration](#classdeclaration)
  * [CompilationUnit](#compilationunit)
  * [Expression](#expression)
  * [FieldAccess](#fieldaccess)
  * [Identifier](#identifier)
  * [MethodDeclaration](#methoddeclaration)
  * [MethodInvocation](#methodinvocation)
  * [NewClass](#newclass)
  * [Statement](#statement)
  * [VariableDeclarations](#variabledeclarations)
* [Teach you how to use the debugger to learn more about ASTs yourself](#using-the-debugger-to-detect-asts)

## High-level AST Explanation

In order for a computer to parse and change code, there needs to be some data structure defined that can accurately represent said code. OpenRewrite uses an [Abstract Syntax Tree](/v1beta/abstract-syntax-trees.md) (AST) for this purpose. Like many tree data structures, ASTs are often composed of many other ASTs. 

> For instance, you could have a [CompilationUnit](#compilationunit) which then has a [ClassDeclaration](#classdeclaration) inside of that which then has a [MethodDeclaration](#methoddeclaration) inside of that and so on. On top of being a valid AST that would _also_ represent valid Java code. 
> 
> On the other hand, you _could_ define a [Binary](#binary) with other Binaries inside of that. That would be a valid AST but it **would not** represent valid Java code as just entering `1+2` (a valid Binary) in a Java file would not compile (that code would need to be in an expression which is then in a class to be valid).
> 
> > Every object in the [code section](#code) is an AST in the sense that it implements a Tree interface. These objects, however, need to be combined in an appropriate way to represent valid Java. It is your duty as a recipe author to ensure that the final AST represents valid Java code.

Reviewer: ```suggestion
For instance, a [ClassDeclaration](#classdeclaration) is an AST which defines a class. A typical class declaration will be composed of elements such as fields, methods, constructors, and inner classes. Each of these are represented as their own, smaller AST. So the term "AST" may refer to an entire, complete Java file or just one piece of it.

It is possible to manipulate ASTs to create code which will not compile. OpenRewrite provides some safeguards against grammatically invalid transformations in its type system - you can't replace an import statement with a method declaration, for example. But there are plenty ways ASTs can be valid according to the Java grammar without being valid, compilable programs. For example, nothing prevents you from modifying an AST such that a variable is used before it is defined. It is the responsibility of recipe authors to consider language semantics and the full range of possibilities when making changes. In accordance with the principle of Doing No Harm, always err on the side of leaving code untouched rather than making a risky change. 
```

Readers don't yet know what a `CompilationUnit` or `ClassDeclaration` are, so these terms cannot yet be used to explain related concepts.
A java programmer is likely to infer the meaning of `ClassDeclaration` correctly, but I had no idea that the name for an entire Java file was a "compilation unit" until I worked here.

## Code

Below is a simple Java class whose entire purpose is to demonstrate different types of ASTs. Each of the following sections will highlight different parts of this code to demonstrate which chunks correspond to which AST.

```java
package org.openrewrite;
                
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@AnAnnotation
public class A {

    List<Integer> a = new ArrayList<>();

    int foo() {
        int a = 1 + 2, b = 3, c;
        this.a = this.a.stream()
                .map(it -> it + 1)
                .collect(Collectors.toList());
        return a;
    }
}

@interface AnAnnotation {}
```

### Binary

> A [Binary AST](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L597-L705) is an [Expression](#expression) with a left and right side, separated by an operator. Examples of operators include `+`, `-`, `||`, `&&`, and more.

Reviewer: The line numbers in these links will be wrong as soon as we make changes to `J`. Which aren't constant, but it does happen. Just today I tweaked an entry in `J`. If we can't make more durable links, and I'm not aware of any way to do that, better to link to `J` once without trying to link to its individual components.

Author: I could update the links to use a tag such as: https://github.com/openrewrite/rewrite/blob/v7.33.0/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java which will stay consistent regardless of the changes made. 

Does that seem reasonable to you or would you prefer I just link to the file rather than the individual parts?

![Binary Example](<../.gitbook/assets/Binary.png>)

### Block 

A [Block AST](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L712-L851) is a pair of curly braces and the [Statements](#statement) contained within. Blocks can be nested inside of each other.

![Block Example](<../.gitbook/assets/Block.png>)

### ClassDeclaration

A [ClassDeclaration AST](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L1062-L1336) contains all of the code for any Java class. Please note that a `ClassDeclaration` can be nested inside of another class such as with:

```java
public class A {
    // ...
    private class B {
        // ...
    }
}
```

![ClassDeclaration Example](<../.gitbook/assets/ClassDeclaration.png>)

### CompilationUnit

> A [CompilationUnit AST](https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java#L1342-L1545) is the root of the Java AST. In order for an AST to represent valid Java code, all other elements must be contained inside of this. 

Reviewer: I don't think we need to say "<AST type> AST" for each of these. It is redundant. 

It's like if we were listing breeds of dogs like this:

- Labrador dog
- Schnauzer dog
- Beagle dog
- Dachshund dog
- Corgi dog

Probably didn't need so many dog types to illustrate my point. But look how cute this beagle is:
![image](https://user-images.githubusercontent.com/8549033/206067786-7add3347-569c-4ba3-b1e3-742ed3a74826.png)

Author: I'm okay with making this change -- but I do want to call out that I'm not sure I agree with that. The naming conventions we have don't make it obvious that these things _are_ ASTs. If I saw a section called, `Identifier` and saw that it talked about variable names, I definitely would not assume it's an AST. That being said, I recognize that I'm still learning this so will default to you on this.
